kafka/consumer.go

```diff
@@ -0,0 +1,393 @@
+		completed[partition] = false
+	}
+
+	storeOffsets := func(ts time.Time) {
+		for partition := range c.currentOffsets {
+			offset := atomic.LoadInt64(c.currentOffsets[partition])
+			c.partitionOffset[partition].Set(int(offset))
+			if c.LagMonitor != nil {
+				c.LagMonitor.StoreOffset(partition, offset, ts)
+			}
+			if !completed[partition] && offset >= c.bootTimeOffsets[partition]-1 {
+				if processBacklog != nil {
+					processBacklog.Done()
+				}
+				completed[partition] = true
+				delete(c.bootTimeOffsets, partition)
+				if len(c.bootTimeOffsets) == 0 {
+					c.bootTimeOffsets = nil
+				}
+			}
+
+			_, newest, err := c.consumer.QueryWatermarkOffsets(c.conf.Topics[0], partition, c.conf.MetadataTimeout)
+			if err != nil {
+				log.Error(3, "kafka-consumer: Error when querying for offsets: %s", err)
+			} else {
+				c.partitionLogSize[partition].Set(int(newest))
+			}
+
+			c.partitionOffset[partition].Set(int(offset))
+			if err == nil {
+				lag := int(newest - offset)
+				c.partitionLag[partition].Set(lag)
+				if c.LagMonitor != nil {
+					c.LagMonitor.StoreLag(partition, lag)
+				}
+			}
+		}
+	}
+
+	ticker := time.NewTicker(c.conf.OffsetCommitInterval)
+	for {
+		select {
+		case ts := <-ticker.C:
+			storeOffsets(ts)
+		case <-c.stopChan:
+			storeOffsets(time.Now())
+			return
+		}
+	}
+}
+
+func (c *Consumer) startConsumer() error {
+	var offset confluent.Offset
+	var err error
+	var topicPartitions confluent.TopicPartitions
+	c.currentOffsets = make(map[int32]*int64, len(c.Partitions))
+
+	for i, topic := range c.conf.Topics {
+		for _, partition := range c.Partitions {
+			var currentOffset int64
+			switch c.conf.StartAtOffset {
+			case "oldest":
+				currentOffset, _, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetBeginning), 3, time.Second)
+				if err != nil {
+					return err
+				}
+			case "newest":
+				_, currentOffset, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetEnd), 3, time.Second)
+				if err != nil {
+					return err
+				}
+			default:
+				offsetDuration, err := time.ParseDuration(c.conf.StartAtOffset)
+				if err != nil {
+					return fmt.Errorf("invalid offest format %s: %s", c.conf.StartAtOffset, err)
+				}
+				currentOffset = time.Now().Add(-1*offsetDuration).UnixNano() / int64(time.Millisecond)
+				currentOffset, _, err = c.tryGetOffset(topic, partition, currentOffset, 3, time.Second)
+				if err != nil {
+					return err
+				}
+			}
+
+			offset, err = confluent.NewOffset(currentOffset)
+			if err != nil {
+				return err
+			}
+
+			topicPartitions = append(topicPartitions, confluent.TopicPartition{
+				Topic:     &topic,
+				Partition: partition,
+				Offset:    offset,
+			})
+
+			if i == 0 {
+				c.currentOffsets[partition] = &currentOffset
+			}
+		}
+	}
+
+	return c.consumer.Assign(topicPartitions)
+}
+
+func (c *Consumer) tryGetOffset(topic string, partition int32, offsetI int64, attempts int, sleep time.Duration) (int64, int64, error) {
```

Comment:
why does this return 2 values?

Reply:
changing that. it made sense in an older version of the code, but now not anymore